Generic Sensor: Optimize sensor tests in iframe

Honry opened the pull request:
Make tests simplified and robust by unifying iframe generation
  in javascript and using promise_test

<!-- Reviewable:start -->

<!-- Reviewable:end -->

Honry commented:
@tobie, could you PTAL?

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240664867)

# Firefox (nightly) #
Testing web-platform-tests at revision 781c1192b2e4506c865d21cf2e5dfd355ac5b96c
Using browser at version BuildID 20170607100302; SourceStamp 5801aa478de12a62b2b2982659e787fcc4268d67
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>8 tests ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/accelerometer/Accelerometer.https.html">/accelerometer/Accelerometer.https.html</a></summary>


|       Subtest        | Results |                    Messages                    |
|----------------------|---------|------------------------------------------------|
|                      | OK      |                                                |
| `Accelerometer Test` | FAIL    | `ReferenceError: Accelerometer is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/accelerometer/Accelerometer_insecure_context.html">/accelerometer/Accelerometer_insecure_context.html</a></summary>


|                Subtest                 | Results |                    Messages                    |
|----------------------------------------|---------|------------------------------------------------|
|                                        | OK      |                                                |
| `Accelerometer Test: insecure context` | FAIL    | `ReferenceError: Accelerometer is not defined` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/ambient-light/AmbientLightSensor.https.html">/ambient-light/AmbientLightSensor.https.html</a></summary>


|          Subtest          | Results |                       Messages                      |
|---------------------------|---------|-----------------------------------------------------|
|                           | OK      |                                                     |
| `AmbientLightSensor Test` | FAIL    | `ReferenceError: AmbientLightSensor is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/ambient-light/AmbientLightSensor_insecure_context.html">/ambient-light/AmbientLightSensor_insecure_context.html</a></summary>


|                   Subtest                   | Results |                       Messages                      |
|---------------------------------------------|---------|-----------------------------------------------------|
|                                             | OK      |                                                     |
| `AmbientLightSensor Test: insecure context` | FAIL    | `ReferenceError: AmbientLightSensor is not defined` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/gyroscope/Gyroscope.https.html">/gyroscope/Gyroscope.https.html</a></summary>


|     Subtest      | Results |                  Messages                  |
|------------------|---------|--------------------------------------------|
|                  | OK      |                                            |
| `Gyroscope Test` | FAIL    | `ReferenceError: Gyroscope is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/gyroscope/Gyroscope_insecure_context.html">/gyroscope/Gyroscope_insecure_context.html</a></summary>


|              Subtest               | Results |                  Messages                  |
|------------------------------------|---------|--------------------------------------------|
|                                    | OK      |                                            |
| `Gyroscope Test: insecure context` | FAIL    | `ReferenceError: Gyroscope is not defined` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/magnetometer/Magnetometer.https.html">/magnetometer/Magnetometer.https.html</a></summary>


|       Subtest       | Results |                    Messages                   |
|---------------------|---------|-----------------------------------------------|
|                     | OK      |                                               |
| `Magnetometer Test` | FAIL    | `ReferenceError: Magnetometer is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/magnetometer/Magnetometer_insecure_context.html">/magnetometer/Magnetometer_insecure_context.html</a></summary>


|                Subtest                | Results |                    Messages                   |
|---------------------------------------|---------|-----------------------------------------------|
|                                       | OK      |                                               |
| `Magnetometer Test: insecure context` | FAIL    | `ReferenceError: Magnetometer is not defined` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240664868)

# Sauce (safari) #
Testing web-platform-tests at revision 781c1192b2e4506c865d21cf2e5dfd355ac5b96c
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>8 tests ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/accelerometer/Accelerometer.https.html">/accelerometer/Accelerometer.https.html</a></summary>


|       Subtest        | Results |                       Messages                       |
|----------------------|---------|------------------------------------------------------|
|                      | OK      |                                                      |
| `Accelerometer Test` | FAIL    | `ReferenceError: Can't find variable: Accelerometer` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/accelerometer/Accelerometer_insecure_context.html">/accelerometer/Accelerometer_insecure_context.html</a></summary>


|                Subtest                 | Results |                       Messages                       |
|----------------------------------------|---------|------------------------------------------------------|
|                                        | OK      |                                                      |
| `Accelerometer Test: insecure context` | FAIL    | `ReferenceError: Can't find variable: Accelerometer` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/ambient-light/AmbientLightSensor.https.html">/ambient-light/AmbientLightSensor.https.html</a></summary>


|          Subtest          | Results |                          Messages                         |
|---------------------------|---------|-----------------------------------------------------------|
|                           | OK      |                                                           |
| `AmbientLightSensor Test` | FAIL    | `ReferenceError: Can't find variable: AmbientLightSensor` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/ambient-light/AmbientLightSensor_insecure_context.html">/ambient-light/AmbientLightSensor_insecure_context.html</a></summary>


|                   Subtest                   | Results |                          Messages                         |
|---------------------------------------------|---------|-----------------------------------------------------------|
|                                             | OK      |                                                           |
| `AmbientLightSensor Test: insecure context` | FAIL    | `ReferenceError: Can't find variable: AmbientLightSensor` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/gyroscope/Gyroscope.https.html">/gyroscope/Gyroscope.https.html</a></summary>


|     Subtest      | Results |                     Messages                     |
|------------------|---------|--------------------------------------------------|
|                  | OK      |                                                  |
| `Gyroscope Test` | FAIL    | `ReferenceError: Can't find variable: Gyroscope` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/gyroscope/Gyroscope_insecure_context.html">/gyroscope/Gyroscope_insecure_context.html</a></summary>


|              Subtest               | Results |                     Messages                     |
|------------------------------------|---------|--------------------------------------------------|
|                                    | OK      |                                                  |
| `Gyroscope Test: insecure context` | FAIL    | `ReferenceError: Can't find variable: Gyroscope` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6185/magnetometer/Magnetometer.https.html">/magnetometer/Magnetometer.https.html</a></summary>


|       Subtest       | Results |                       Messages                      |
|---------------------|---------|-----------------------------------------------------|
|                     | OK      |                                                     |
| `Magnetometer Test` | FAIL    | `ReferenceError: Can't find variable: Magnetometer` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6185/magnetometer/Magnetometer_insecure_context.html">/magnetometer/Magnetometer_insecure_context.html</a></summary>


|                Subtest                | Results |                       Messages                      |
|---------------------------------------|---------|-----------------------------------------------------|
|                                       | OK      |                                                     |
| `Magnetometer Test: insecure context` | FAIL    | `ReferenceError: Can't find variable: Magnetometer` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240664874)

# Chrome (unstable) #
Testing web-platform-tests at revision 781c1192b2e4506c865d21cf2e5dfd355ac5b96c
Using browser at version 60.0.3112.20 dev
Starting 10 test iterations

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240664875)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 781c1192b2e4506c865d21cf2e5dfd355ac5b96c

Honry committed:
Generic Sensor: Optimize sensor tests in iframe

  Make tests simplified and robust by unifying iframe generation
  in javascript and using promise_test

Honry commented:
@alexshalamov or @pozdnyakov, PTAL.

alexshalamov commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -32,6 +32,40 @@ function reading_to_array(sensor) {
+
+function wait_for_message(iframe) {
+  return new Promise(resolve => {
+    self.addEventListener("message", function listener(e) {
```
Why not to use window.onmessage here?

Honry replied:
This optimization refers to [`idbfactory-open-opaque-origin.html`](https://github.com/w3c/web-platform-tests/blob/master/IndexedDB/idbfactory-open-opaque-origin.html), I think the idea is to clean up the listener once it receives the message to avoid secondary listening.

alexshalamov commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -141,11 +175,15 @@ function runGenericSensorTests(sensorType) {
+  promise_test(t => {
+    return load_iframe(make_script(sensorType))
+      .then(iframe => {
+        iframe.contentWindow.postMessage({}, '*');
```
Could there be a race condition? E.g., reply is received from iframe's postMessage before parent is subscribed. Also, this test fails with timeout error.

Honry replied:
```promise_tests``` run sequentially, so this should not be the case.

alexshalamov replied:
I was referring to a different problem, first message is posted to an iframe so that it can try constructing sensor object and reply back, then Promise starts listening to message. I'm not sure whether reply from the iframe can be lost.

> Also, this test fails with timeout error.

Do you have same issue with the test?

Honry replied:
Got you point. I didn't find any evidence doc but with some testing on browsers, it works fine that calling postMessage before message listener created.

alexshalamov replied:
@Honry just tried https://w3c-test.org/submissions/6185/accelerometer/Accelerometer.https.html iframe test is failing with Timeout error. Tested with Chrome canary on Android.

Honry replied:
@alexshalamov, seems there's a bug in ```submisssions``` wpt server that it references the ```/generic-sensor/generic-sensor-tests.js``` from https://w3c-test.org/generic-sensor/generic-sensor-tests.js rather than  https://w3c-test.org/submissions/6185/generic-sensor/generic-sensor-tests.js.
It works fine on my personal repo, https://honry.github.io/demo/accelerometer/Accelerometer.https.html

alexshalamov commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -32,6 +32,40 @@ function reading_to_array(sensor) {
+    iframe.onload = () => { resolve(iframe); };
+    iframe.srcdoc = src;
+    iframe.style.display = "none";
+    document.documentElement.appendChild(iframe);
```
document.body?

Honry replied:
Actually they could accomplish the same goal, but yes, ```document.body``` is more accurate.

pozdnyakov commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -32,6 +32,40 @@ function reading_to_array(sensor) {
+  });
+}
+
+function make_script(sensorType) {
```
Could the name of this function be more descriptive? tryToCreateSensorObject or smth else?

Honry replied:
Sure, how about “try_to_create_sensor”, use lower lowercase to keep consistent.

alexshalamov reviewed (changes requested):
Changes look fine, however, test needs to be working before PR is merged.

alexshalamov commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -141,11 +175,15 @@ function runGenericSensorTests(sensorType) {
+      .then(message => {
+        assert_equals(message.result, 'SecurityError');
+      });
   }, "throw a 'SecurityError' when firing sensor readings within iframes");
```
> throw a 'SecurityError' when firing sensor readings within iframes

throw a 'SecurityError' when constructing sensor object within iframe

Honry replied:
thanks, fixed.

Honry committed:
Use a descriptive function name

 -use document.body instead of document.documentElement

Honry committed:
Improve test desc

alexshalamov commented:
@Honry I added layout test for [Sensors](https://chromium-review.googlesource.com/c/559044/), similar to [NFC](https://cs.chromium.org/chromium/src/third_party/WebKit/LayoutTests/nfc/nfc-block-iframe.html), but simpler.

Honry commented:
Yeah, I saw it, will rework this PR later.

Honry committed:
make test simpler

Honry committed:
make test simpler

Honry commented:
@alexshalamov, please check again.

alexshalamov commented:
@Honry Sorry, I was OOO. Is it possible to trigger w3c-bots for last changes?

Honry commented:
@alexshalamov, I think it is possible but I don't know how. :(

alexshalamov reviewed (approved):
test is passing now, thanks

Honry commented:
@alexshalamov, thanks for your review, I will merge it directly.
